server/api/controllers/image-controller.ts

```diff
@@ -0,0 +1,271 @@
+        res.status(403).json({ status: "fail", message: foundFiles });
+        return;
+      }
+
+      unlinkSync(path);
+      this.logger.info(`${fileName} deleted from ${path}`);
+      this.rewriteImageJson();
+      res.status(200).json({ status: "success" });
+    } catch (err) {
+      this.logger.error(err);
+      res.status(400).json({ status: "fail" });
+    }
+  }
+
+  private static generateJSON(): Image[] {
+    const uploadedFiles: Dirent[] = readdirSync(this.getImagePath(""), {
+      withFileTypes: true,
+    });
+
+    const images: Image[] = [];
+
+    for (const img of uploadedFiles) {
+      if (img.isFile()) {
+        images.push({
+          fileName: img.name,
+          fileType: img.name.split(".")[1],
+          path: this.getImagePath(img.name),
+          publicLink: this.getPublicLink(img.name),
+        });
+      }
+    }
+
+    return images;
+  }
+
+  private static rewriteImageJson(): void {
+    const url = "_hidden/image-list";
+    ReadWriteController.overwriteJSONDataPath(
+      url,
+      () => {
+        return;
+      },
+      this.generateJSON()
+    );
+    this.logger.info("Image list regenerated.");
+  }
+
+  private static getImagePath(fileName: string) {
+    return `public/assets/img/uploads/${fileName}`;
+  }
+
+  private static getPublicLink(fileName: string) {
+    return `/assets/img/uploads/${fileName}`;
+  }
+
+  // find occurences of image inside data folder before deleting.
+  private static getDataFiles(mainPath: string, target: string): string[] {
+    const files: string[] = [];
+
+    const dataDir = readdirSync(mainPath);
+    dataDir.forEach((file) => {
+      const filePath = path.join(mainPath, file);
+      const stat = lstatSync(filePath);
+
+      if (stat.isDirectory()) {
+        if (!filePath.includes("_hidden")) {
+          const nestedFiles = this.getDataFiles(filePath, target);
+          nestedFiles.forEach((file) => files.push(file));
+        }
+      } else {
+        files.push(filePath);
+      }
+    });
+
+    return files;
+  }
+
+  private static checkImageOccurences(mainPath: string, target: string) {
+    const files = this.getDataFiles(mainPath, target);
+    const matches: string[] = [];
+    for (const file of files) {
+      const fileContent = readFileSync(file);
+      if (fileContent.includes(target)) {
+        matches.push(file);
+      }
+    }
+
+    return matches;
+  }
+}
+
+class ImageRequestQueue {
+  private queue: (ImageUploadRequest | ImageDeleteRequest)[] = [];
+
+  push(
+    img: Image,
+    action: Action,
+    res: Response,
+    uploadedFile?: UploadedFile
+  ): void {
+    const data: ImageUploadRequest | ImageDeleteRequest = {
+      fileName: img.fileName,
+      fileType: img.fileType,
+      path: img.path,
+      publicLink: img.publicLink,
+      type: action,
+      uploadedFile: uploadedFile ?? null,
+      res,
+    };
+
+    this.queue.push(data);
+
+    if (this.queue.length === 1) {
+      this.process();
+    }
```

Comment:
is this if statement necessary? I feel like I'm missing something 🤔

Reply:
Yea, it starts processing the requests in the image queues since   push is how we start the process queue. Then in the process function, if there are more requests left the process, it calls itself again (line 269)